app/data_sources/gitlab.py

```diff
@@ -0,0 +1,107 @@
+from datetime import datetime
+
+import requests
+import concurrent.futures
+
+from typing import Dict
+
+from indexing_queue import IndexingQueue
+from data_source_api.basic_document import BasicDocument, DocumentType
+from data_source_api.base_data_source import BaseDataSource
+from data_source_api.exception import InvalidDataSourceConfig
+from pydantic import BaseModel
+
+GITLAB_BASE_URL = "https://gitlab.com/api/v4"
+PROJECTS_URL = f"{GITLAB_BASE_URL}/projects?membership=true"
+
+
+class GitlabConfig(BaseModel):
+    access_token: str
+
+
+class GitlabDataSource(BaseDataSource):
+    @staticmethod
+    def validate_config(config: Dict) -> None:
+        try:
+            parsed_config = GitlabConfig(**config)
+            session = requests.Session()
+            session.headers.update({"PRIVATE-TOKEN": parsed_config.access_token})
+            projects_response = session.get(PROJECTS_URL)
+            if projects_response.status_code != 200:
+                raise ValueError("Invalid api key")
+        except (KeyError, ValueError) as e:
+            raise InvalidDataSourceConfig from e
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Create a access token with sufficient permissions in https://gitlab.com/-/profile/personal_access_tokens
+        self.gitlab_config = GitlabConfig(**self._config)
+        self.session = requests.Session()
```

Comment:
Please follow PEP8 guidelines and change to `_session`, thanks.

Reply:
Done